prov/tcp: Msg Xfer fixes, optimizations

vkrishna opened the pull request:
First commit is about adding locking around ep tx/rx queues.
Second commit restructures progress so that disconnect is detected even when queues are empty.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -173,35 +173,50 @@ static void process_rx_pe_entry(struct tcpx_pe_entry *pe_entry)
-	process_pe_lists(ep);
+
+	progress_tx_queue(ep);
+	progress_rx_queue(ep);
```
Both of these functions acquire and release the same lock.  Can the locking be moved here and held throughout?

shefty commented on prov/tcp/src/tcpx.h:
```diff
@@ -188,9 +188,10 @@ struct tcpx_pep {
 	struct dlist_entry	rx_queue;
 	struct dlist_entry	tx_queue;
+	/* lock for accessing rx & tx queues */
+	fastlock_t		queue_lock;
```
The udp provider ends up using the tx/rx cq locks when sending or receiving.  This avoids adding extra locks when a completion must also be written.  Please see if a similar model would be useful here.

vkrishna replied:
I think it is good idea to use the cq locks.  That will decrease the number of locks for sure. But I am wondering if it would add considerable performance overhead.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -122,29 +122,20 @@ static ssize_t tcpx_sendmsg(struct fid_ep *ep, const struct fi_msg *msg,
-	if (flags & FI_INJECT) {
-		assert(data_len <= TCPX_MAX_INJECT_SZ);
-	}
+	assert((flags & FI_INJECT)?data_len <= TCPX_MAX_INJECT_SZ:1);
```
This can be converted to:
```
assert(!(flags & FI_INJECT) || (data_len <= TCPX_MAX_INJECT_SZ));
```

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
+	int ret, i;
+
+	fastlock_acquire(&mgr->lock);
+	ret = fi_epoll_wait(mgr->epoll_fd, mgr->ctxs,
```
We don't want to enter a blocking call holding a spinlock.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
+		progress_tx_queue(ep);
+	}
+	fastlock_release(&mgr->lock);
+}
```
Maybe this function is misnamed.  I don't see how it does auto progress.  It just tries to progress all endpoints.

vkrishna replied:
Yes I will fix it

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
+	mgr->used--;
+out:
+	fastlock_release(&mgr->lock);
+}
```
ep_add allows inserting the same ep multiple times, but doesn't provide reference counting.  The result is that the first call to ep_del will remove the ep.

vkrishna replied:
added a reference count.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
+			      util_domain);
+
+	tcpx_auto_prog_tx_queues(&domain->progress.mgr);
+	tcpx_auto_prog_rx_queues(&domain->progress.mgr);
```
Manual progress should really be associated with a CQ.  The utility code has callbacks for this, including linking into wait sets and poll sets.

vkrishna replied:
I will make the changes aptly.

shefty commented on prov/tcp/src/tcpx.h:
```diff
@@ -187,13 +188,19 @@ struct tcpx_pep {
 	struct dlist_entry	tx_queue;
 	/* lock for accessing rx & tx queues */
 	fastlock_t		queue_lock;
+	struct tcpx_cm_state	cm_state;
```
Just embed the 2 fields directly into this structure.  If a state is needed, a generic state or set of flags would be preferable than a simply boolean to track whether shutdown has been reported.

vkrishna commented:
@shefty I've moved the entry memory pool from domain to cq in favor of using cq lock for alloc and release. Also added queue lock to protect rx/tx queues of ep.

shefty commented on prov/tcp/src/tcpx_cq.c:
```diff
@@ -35,19 +35,66 @@
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,"failed to get buffer\n");
+		return NULL;
+	}
+	memset(pe_entry, 0, sizeof(*pe_entry));
```
This is cleared both on acquire and in the release call.

vkrishna replied:
removed in release

shefty commented on prov/tcp/src/tcpx_cq.c:
```diff
@@ -35,19 +35,66 @@
+	fastlock_acquire(&cq->cq_lock);
+	memset(&pe_entry->msg_hdr, 0, sizeof(pe_entry->msg_hdr));
+	dlist_remove(&pe_entry->entry);
+	memset(pe_entry, 0, sizeof(*pe_entry));
```
We should look at avoiding blanket memset calls and only initialize whatever fields must be cleared.  If we initialize the header immediately after getting an entry, we can avoid the memset.  Likewise, do all pe_entry fields need to be 0?

shefty commented on prov/tcp/src/tcpx_cq.c:
```diff
@@ -83,20 +130,36 @@ int tcpx_cq_open(struct fid_domain *domain, struct fi_cq_attr *attr,
+		attr->size = TCPX_DEF_CQ_SIZE;
+
+	ofi_atomic_initialize64(&tcpx_cq->cq_free_size,
+				attr->size);
```
I don't see that this value is used anywhere in this patch.

vkrishna replied:
I will remove in this patch and put it in apt patch.

shefty commented on prov/tcp/src/tcpx_conn_mgr.c:
```diff
@@ -260,6 +260,15 @@ static int send_conn_req(struct poll_fd_mgr *poll_mgr,
+static int tcpx_ep_msg_xfer_enable(struct tcpx_ep *ep)
+{
+	fastlock_acquire(&ep->cm_state_lock);
+	ep->cm_state = TCPX_EP_CONN_ACTIVE;
```
Do we need to check that we're in the correct state here before changing it?  E.g. could the ep be disconnected for some reason prior to the user attempting to activate it?

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -391,12 +390,12 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
+		return ret;
 	}
+
 	return FI_SUCCESS;
```
nit: could just fall through and return ret in all cases.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -42,6 +42,42 @@
+
+	fastlock_acquire(&ep->cm_state_lock);
+	if (ep->cm_state == TCPX_EP_CONN_SHUTDOWN)
+		goto success;
```
nit: we can go to 'err' label in this case, as the state is already correct.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -42,6 +42,42 @@
+			ret = (int) -errno;
+			goto err;
+		}
+		goto success;
```
restructure to remove goto.  Place code after if (err) check above into else clause.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -42,6 +42,42 @@
+				&err_entry, sizeof(err_entry),
+				UTIL_FLAG_ERROR) < 0) {
+			ret = (int) -errno;
+			goto err;
```
We should probably still set the state to shutdown even if we can't write an event to the eq.

shefty replied:
I believe all CM events should use fi_eq_cm_entry

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -42,6 +42,42 @@
+	}
+success:
+	ep->cm_state = TCPX_EP_CONN_SHUTDOWN;
+err:
```
rename label to 'unlock' or 'out', since non-error flows go to here.

shefty commented on prov/tcp/src/tcpx.h:
```diff
@@ -165,6 +165,12 @@ struct tcpx_pep {
+	TCPX_EP_INIT,
+	TCPX_EP_CONN_ACTIVE,
+	TCPX_EP_CONN_SHUTDOWN,
+};
```
We may want to have these states: CONNECTING, CONNECTED, SHUTDOWN, ERROR.  These are really active endpoint states, not necessarily CM states.  You can probably drop the 'CONN' from the name.

vkrishna replied:
updated

vkrishna commented:
Thanks Sean.

vkrishna commented:
@shefty Added the fast path for small messages with the last commit.

shefty commented on prov/tcp/src/tcpx_cq.c:
```diff
@@ -57,6 +57,12 @@ struct tcpx_pe_entry *tcpx_pe_entry_alloc(struct tcpx_cq *tcpx_cq)
+
+	/* optimization: don't allocate queue_entry when cq is full */
+	if (ofi_cirque_isfull(tcpx_cq->util_cq.cirq)) {
+		return NULL;
```
This returns while holding a lock

vkrishna replied:
Fixed it.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -170,7 +170,13 @@ static ssize_t tcpx_sendmsg(struct fid_ep *ep, const struct fi_msg *msg,
+		process_tx_pe_entry(send_entry);
+	} else {
+		dlist_insert_tail(&send_entry->entry, &tcpx_ep->tx_queue);
+	}
```
The 'fast' path here is based on the tx_queue being empty, and not on message size.

vkrishna commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -129,3 +154,39 @@ void tcpx_progress(struct util_ep *util_ep)
+		fastlock_release(&ep->cm_state_lock);
+		return;
+	}
+	fastlock_release(&ep->cm_state_lock);
```
When the state is still connecting, we just bail out skipping ep delete. Once we are in some other state, we won't go back to that state again. So state going back to "TCPX_EP_CONNECTING" after releasing lock would not be an issue.

shefty replied:
It's not clear what the lock is protecting against is the main issue.  The state could change to CONNECTING immediately after releasing the lock, for example.

vkrishna replied:
I have added this check to prevent calling ep_del before ep_add ( i.e., before the connection was established).  

But since the ep_del gets called during ep close and conn management happens in another thread, there might be a race condition where ep_close might get the lock first and ep_add later. Thus ending up with closed socket in cq->wait object.

shefty commented on prov/tcp/src/tcpx_progress.c:
```diff
@@ -129,3 +154,39 @@ void tcpx_progress(struct util_ep *util_ep)
+		if (ofi_atomic_dec32(&ep->ref_cnt))
+			return;
+
+		ofi_wait_fd_del(ep->util_ep.rx_cq->wait, ep->conn_fd);
```
Yes, I think you can drop your reference counting.
